File: 백준/Gold/1238. 파티/파티.java
import java.io.*;
import java.util.*;


public class Main {
	public int[] dijkstra(List<List<int[]>> tree, int X, int N) throws Exception {
		int[] dists = new int[N+1];
		Arrays.setAll(dists, i->Integer.MAX_VALUE);
		dists[X] = 0;
		
		PriorityQueue<int[]> hq = new PriorityQueue<int[]>((a, b)->a[1]-b[1]);
		hq.add(new int[] {X, 0});
		while (!hq.isEmpty()) {
			int[] curArr = hq.poll();
			int curNode = curArr[0];
			int curDist = curArr[1];
			
			if (dists[curNode]>curDist) continue;
			
			for (int[] nextArr : tree.get(curNode)) {
				int nextNode = nextArr[0];
				int nextDist = curDist+nextArr[1];
				
				if (nextDist<dists[nextNode]) {
					dists[nextNode] = nextDist;
					hq.add(new int[] {nextNode, nextDist});
				}
			}
		}
		
		return dists;
	}
	
	
	public void solution() throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        StringTokenizer st = new StringTokenizer(br.readLine());
        
        int N, M, X;
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());
        X = Integer.parseInt(st.nextToken());

        List<List<int[]>> tree = new ArrayList<>();
        List<List<int[]>> reversedTree = new ArrayList<>();
        for (int i=0;i<N+1;i++) {
        	tree.add(new ArrayList<>());
        	reversedTree.add(new ArrayList<>());
        }
        
        for (int i=0;i<M;i++) {
        	st = new StringTokenizer(br.readLine());
        	int s, e, t;
        	s = Integer.parseInt(st.nextToken());
        	e = Integer.parseInt(st.nextToken());
        	t = Integer.parseInt(st.nextToken());
        	
        	tree.get(s).add(new int[] {e, t});
        	reversedTree.get(e).add(new int[] {s, t});
        }
        
        int[] fromX = dijkstra(tree, X, N);
        int[] toX = dijkstra(reversedTree, X, N);
        int[] totalDist = new int[N+1];
        Arrays.setAll(totalDist, i->i==0 ? 0 : fromX[i]+toX[i]);
        
        bw.write(Arrays.stream(totalDist).max().getAsInt()+"");
        br.close();
        bw.flush();
        bw.close();
    }
    
    
    public static void main(String[] args) throws Exception {
        new Main().solution();
    }
}

File: 백준/Gold/1238. 파티/파티.py
import heapq
import sys

input = sys.stdin.readline


def dijkstra(graph, start):
    min_dists = [float('inf')] * len(graph)
    min_dists[start] = 0

    hq = [[0, start]]
    while hq:
        cur_dist, cur_node = heapq.heappop(hq)

        if min_dists[cur_node] < cur_dist:
            continue

        for next_node, dist in graph[cur_node]:
            next_dist = cur_dist + dist

            if next_dist < min_dists[next_node]:
                heapq.heappush(hq, [next_dist, next_node])
                min_dists[next_node] = next_dist

    return min_dists


N, M, X = map(int, input().split())
graph = [[] for _ in range(N + 1)]
reverse_graph = [[] for _ in range(N + 1)]

for _ in range(M):
    a, b, t = map(int, input().split())
    graph[a].append([b, t])
    reverse_graph[b].append([a, t])

total_dists = [t1 + t2 for t1, t2 in zip(dijkstra(graph, X), dijkstra(reverse_graph, X))][1:]
print(max(total_dists))